### ISSUE TITLE
[HL2DM] 22KHz sounds are not playing correctly client side

### ISSUE THREAD

**Issue by henk717** [ValveSoftware/Source-1-Games#6620]
The anniversary update has severely broken the sounds, this is noticeable on various effects such as the charging sound in various maps but this also happens on RPG missiles. Because my server runs on the old SDK2013 version with a compatibility patch I can confirm this issue happens client side as the behavior is identical on both the old SDK2013 based server, beta based server and listen server. 

Sounds are cut off and distorted. All players have this issue.

**Comment by henk717** [comment]
Easy reproduction steps:
1. Open HL2DM.
2. Load dm_lockdown with create servers.
3. Use the super charger, audio will be silent until finished then it will play briefly in low quality.

**Comment by henk717** [comment]
We have isolated this issue to sounds sampled at 22khz of which Half-Life 2 has multiple important sounds and this is a critical bug as a result. Players can not correctly hear RPG missiles, tripwire sounds and charging sounds.

**Comment by nicholas-castro** [comment]
This is a severe bug, are there any plans to fix this?

**Comment by yurakone** [comment]
https://github.com/yurakone/hl2dm_soundfix

**Comment by adelina1985a** [comment]
I'm seeing fixes or attempts for a fix in parallel in other venues like different discord communities and while I would love to have everyone combine forces, all I can do is suggest that they use this as the main platform for discussing and finding solutions. And I ask people here to do the same.

Regarding one solution I read: they were reducing the pitch of the sounds, BUT using sourcemod which not all servers have or want. This seems to work as an alternative to changing the sample rate, so why not modify the pitch in the soundscripts? scripts\game_sounds_blah.txt.
I don't know which would be better in terms of sample rate vs pitch modification, but a cleaner solution would be to not have the client download extra sounds. And a universal solution that can be applied to ALL servers would be even better. And of course a fix from the devs would be the best.

Here's a list of sounds I got when I think I deleted my sound cache... I think, idk, I tried a lot of things!
```
Cannot figure out which search path sound\server_sounds\smallmedkit1.wav came from. Not playing.
Cannot figure out which search path sound\server_sounds\mine_activate.wav came from. Not playing.
Cannot figure out which search path sound\server_sounds\rocket1.wav came from. Not playing.
Cannot figure out which search path sound\server_sounds\sniper_zoomin.wav came from. Not playing.
Cannot figure out which search path sound\server_sounds\sniper_zoomout.wav came from. Not playing.
Cannot figure out which search path sound\server_sounds\medshotno1.wav came from. Not playing.
Cannot figure out which search path sound\server_sounds\medshot4.wav came from. Not playing.
Cannot figure out which search path sound\server_sounds\medcharge4.wav came from. Not playing.
Cannot figure out which search path sound\server_sounds\suitchargeno1.wav came from. Not playing.
Cannot figure out which search path sound\server_sounds\suitchargeok1.wav came from. Not playing.
Cannot figure out which search path sound\server_sounds\suitcharge1.wav came from. Not playing.
```
This list is **not complete**. While ammo crates are not an issue for me, I am also audibly missing 
`common\wpn_select.wav` before charging, and the client side only weapon scroll sounds:
`common\wpn_moveselect.wav`
`common\wpn_hudoff.wav`
How I got the names? See my comment here https://github.com/ValveSoftware/Source-1-Games/issues/6570
Why am I listing all this? Because adding
`"pitch"		"99"`
to these sounds works, I tested some of them. The weapon scroll ones are in `game_sounds_ui.txt` (I took it from `C:\Program Files (x86)\Steam\steamapps\common\Half-Life 2 Deathmatch\hl2\scripts`), just modify that and put it in your custom\bla\scripts. For me it works on updated servers! The wpn_select is in `game_sounds_player.txt`, I modified only the first instance and it works on local only as expected, because pretty much all non-ui sounds will have to be a server side fix. The medshotno1 is in `game_sounds_items.txt`, also works in local only (and on servers that have their variant of a fix). GL&HF!

EDIT: I was unable to test the following sound due to not first spawning on both sides of dm_lockdown (https://github.com/ValveSoftware/Source-1-Games/issues/6760) plus it stops short in the start of the map:
`npc\attack_helicopter\aheli_wash_loop3.wav`
If people can remember, it sounds like the fan. Also these would be temp fixes while we wait for the real deal...

EDIT2: More sounds, these ones go in game_sound_weapons.txt and I had to copy the soundscript entries from garrysmod, for some reason they are missing in my txt version:
```
weapons/slam/throw.wav
weapons/slam/mine_mode.wav
weapons/tripwire/mine_activate.wav
```
A client side fix (pitch 99) made the first 2 work on servers as well. The client side fix for the 3rd only works in local. There's 2 more slam sounds, for placing and detonating, but they already work :)

**Comment by adelina1985a** [comment]
@henk717 Peter Brev said you identified over a thousand sounds, which is a lot more than Salo Power has a modified sample rate for, and a lot more than the ones I found so far. I think we can ignore the map specific ones for now, but do you have a list that you can post or link here? And mention which ones you got to test/compare against legacy and confirm that they are broken, please.

**Comment by henk717** [comment]
Thats outdated info now. There are over 1000 sounds that are 22khz but on further inspection not all of them are broken so that was misleading and based on early findings as converting those to 41khz improves the  situation usually. The current most practial ways of working around the bug is twofold. As a server operator you can install a special workaround plugin that changes the pitch from 100 to 99 for all pitch 100 sounds, audibly this is no different since its to close to be perceived. Players can choose to play on these servers and if their favorite server does not have this fix they can petition the server owner to install https://github.com/Alienmario/hl2dm-anniversary-soundfix .

As a player you can also upsample all the sounds that personally effect you or upsample all of them in builk (wavegold is a good tool for this). Now sounds will play in every server that is not a sv_pure server. The issue we can't get around is that some of the sounds play at a noticably low bitrate even if the original plays fine. The AR2 is an example of a 22khz sound that turns out to work fine. But if upsampled sounds low bitrate for no apparent reason. So then if you encounter those sounds that are otherwise fine you'd have to delete them.

All in all valve should just fix the bug with unprocessed sound playback. But in their absence the plugin is the most viable. You then only have to patch clientside sounds like the weapon scrolling and tripmine mode switch.

Update: I saw your earlier comment about not wanting sourcemod fixes and wanting to do it trough config files instead. If a server can get away with this depends on the maps they run. As long as you can edit every sound in the config your good, my servers cant use this because some of the maps have the issue baked in. So for me the sourcemod fix is a hassle free way to do it. The pitch shifting can also be done in SDK2013 of course you don't have to use sourcemod if you are an advanced server and speedvoltage has already done so in his code. So if the objection to sourcemod is that sourcemod runs poorly due to a custom SDK2013 just grab his fix.

**Comment by adelina1985a** [comment]
@henk717 Thank you, that's awesome news that not all 1000 sounds of 22khz are broken, the list might actually be manageable.

"Now sounds will play in every server that is not a sv_pure server." - wouldn't pitch 99 in soundscripts fix that for all servers regardless of sv_pure value? As opposed to clients changing the sample for each troublesome .wav file and having it play only on some servers? It sounds like you support my clientside solution :D

"I saw your earlier comment about not wanting sourcemod fixes and wanting to do it trough config files instead. If a server can get away with this [...]" - Yes, it will be up to the server owner, it's good to give everyone choices. Because what about those that can't or won't do smod/custom code? The existence of vanilla servers is not a personal opinion. Also I have used sourcemod and supported alliedmods for many many years. And even for owners that want sourcemod, ~1/10th of GSPs will need a ticket just to install sourcemod and then another ticket each time there's an update or you want to add a specific plugin/extension. And custom code is a whole different thing that v few will want to get into, either via VPS and setting up everything themselves, or Dedi for a fleet of servers. In general I am not against that either, IF it's modularized. In particular, speedvoltage's repo looks like it had a bunch of fixes merged into the 'essentials' **besides** the sound fix?... And one server in particular that has 'fixed binaries' in the title made all my footsteps silent when I launched local right after. Yikes.
So was just hoping for a solution that **any** server can afford/apply, including those that can't or won't do smod/custom code. And the one I proposed is starting with sounds that can be played regardless of map. It's also easier to test, as mentioned. Just edited my post and added some slam sounds that are fixable that way, and am asking people to post any list of sounds they can find.

### Adding to the list of sound issues, possibly related:
- firing smg alt fire will play the sound (which oddly enough is called ar2_altfire) if you have 2 or 3 to begin with, but when you are down to 1, it's cut off by the 'empty' sound; now... I got it to work with cheat host_timescale 0.3 in _local_ (I'm so tired of using this word that I might buy a server now) but since the sound DOES play in the scenarios mentioned and since this value doesn't fix the broken sounds, it could be a different sound issue entirely, or indicative of the need of a major sound system rehaul;
- a player starting voice chat will suddenly make 1 or more of the otherwise inaudible sounds be audible at a higher volume (possibly those that recently failed to play and got buffered?)
- some sounds (spawns, fights, deaths) that have a source visibly far away, or in the next room/floor, play louder than they should and seem closer than they are.
- **EDIT**: adding that the '_Silence the audio while the game window is in the background_' option is broken, I was alt+tabbed when I heard the ringing from an explosion, more proof that there's a sound buffer issue.

**Comment by speedvoltage** [comment]
> In particular, speedvoltage's repo looks like it had a bunch of fixes merged into the 'essentials' besides the sound fix?

Because the essential fix repo was created for one person who requested bug fixes only. It never was my main repo, but the sound fix is planned in it when I get to it.
> And one server in particular that has 'fixed binaries' in the title made all my footsteps silent when I launched local right after. Yikes.

This happens because footstep sounds are moved to the server side and no longer plays client-side to patch the footstep sounds being bugged as well. If you use `map <mapname>` in the console to boot into a listen server, you also carry all the ConVars from the previous server apparently, which includes `sv_footsteps 0`, which you cannot modify without custom binaries because it is protected by the `FCVAR_DEVELOPMENTONLY` flag. You would need to use `Create Server` to reset all the ConVars again.

**Comment by adelina1985a** [comment]
> If you use `map <mapname>` in the console to boot into a listen server, you also carry all the ConVars from the previous server apparently, which includes `sv_footsteps 0`.

I was looking at `mp_footsteps`. But using `sv_footsteps 1` in listenserver.cfg worked even when using the map command, thanks.
~~However that server in particular did a CTD after changing the spray via the Options Menu. Other servers I tested this on did not crash this way, if at all. Here is the Engine Error (which usually only happens when using snd_restart without the server side fix):
![image](https://github.com/user-attachments/assets/4e9dfa3d-5c64-4b79-bc8e-bb34e85d77cd)~~
Scratch that, I found another server (it's using a previous build, but not legacy) missing the fix and it also crashes when using the Options Menu.

Anyway, the modularity of such a project is still a concern, which I won't discuss here any further. But I do believe it's good to give the server owner some options! Let's leave it at that :)

**Comment by Xeogin** [comment]
> EDIT2: More sounds, these ones go in game_sound_weapons.txt and I had to copy the soundscript entries from garrysmod, for some reason they are missing in my txt version:
> 
> ```
> weapons/slam/throw.wav
> weapons/slam/mine_mode.wav
> weapons/tripwire/mine_activate.wav
> ```
> 
> A client side fix (pitch 99) made the first 2 work on servers as well. The client side fix for the 3rd only works in local. There's 2 more slam sounds, for placing and detonating, but they already work :)

There's another game_sounds_weapons.txt in the hl2mp_pak_dir.vpk that has the SLAM entries

**Comment by adelina1985a** [comment]
@Xeogin Oh, guess I stopped checking that for ever wav, since that vpk only has 2 soundscripts. Thank you. It makes sense that that vpk references some sounds found only in hl2dm.

**Comment by speedvoltage** [comment]
>  But using sv_footsteps 1 in listenserver.cfg worked even when using the map command, thanks.

Looks like many of the commands that were previously behind FCVAR_DEVELOPMENTONLY are unlocked, including sv_footsteps, which is why it worked after all.

**Comment by RiverHornet3209** [comment]
This bug is not fixed on the 32-bit version of HL2DM (hl2mp.exe):

<sup>* Don't forget to unmute the audio.</sup>

https://github.com/user-attachments/assets/dbfb9646-6365-4bb6-93f7-e3ece2c400c0

Entering the `snd_restart` command in the console doesn't fix the issue

**Comment by RiverHornet3209** [comment]
The sounds are also broken on the 32-bit versions of CS:S, HLDM:S, TF2 and DOD:S

<sup>* Don't forget to unmute the audio.</sup>

https://github.com/user-attachments/assets/4ff69709-a352-4d10-bbde-79d8be907cbe

https://github.com/user-attachments/assets/5b109e93-ba5f-408f-977c-4970091bd064

https://github.com/user-attachments/assets/7d9a4541-f349-483c-8a4b-bb7ffbd6ccb4

https://github.com/user-attachments/assets/661c4340-6d03-41b0-b1d9-3f1815155ce6

and also, sprayer.wav has a 11025hz sample rate so this issue also affects certain 11025hz sample rate sounds

**Comment by RiverHornet3209** [comment]
the sound bug also occurs on the 32-bit version of TF2:

https://github.com/user-attachments/assets/d9623f45-81de-432d-8aba-2972cfb55890
for example, the weapon selection sounds aren't working

**Comment by RiverHornet3209** [comment]
this bug also occurs on the 32-bit version of DOD:S:

https://github.com/user-attachments/assets/661c4340-6d03-41b0-b1d9-3f1815155ce6

for example, the spray sound isn't working on 32-bit

**Comment by Adrianilloo** [comment]
Is the issue fixed under 64 bits with the latest updates? Wondering about #3202 as well

**Comment by RiverHornet3209** [comment]
> Is the issue fixed under 64 bits with the latest updates?

It's fixed in 64-bit but not fixed in 32-bit
